pkg/blockbuilder/blockbuilder_test.go

```diff
@@ -473,9 +474,54 @@ func TestBlockBuilder_StartupWithExistingCommit(t *testing.T) {
 	)
 }
 
+// Testing block builder starting up with an existing kafka commit.
+func BenchmarkBlockBuilder(b *testing.B) {
+	ctx, cancel := context.WithCancelCause(context.Background())
+	b.Cleanup(func() { cancel(errors.New("test done")) })
+
+	_, addr := testkafka.CreateClusterWithoutCustomConsumerGroupsSupport(b, 1, testTopic)
+	writeClient := newKafkaProduceClient(b, addr)
+
+	cfg, overrides := blockBuilderConfig(b, addr)
+
+	// Producing some records
+	cycleEnd := cycleEndAtStartup(cfg.ConsumeInterval, cfg.ConsumeIntervalBuffer)
+	kafkaTime := cycleEnd.Truncate(cfg.ConsumeInterval).Add(-1 * time.Hour)
+	var expSamples []mimirpb.Sample
+	for i := int64(0); i < int64(time.Hour/(1*time.Second)); i++ {
+		kafkaTime = kafkaTime.Add(1 * time.Second)
+		sampleTs := kafkaTime.Add(-time.Minute)
+		samples := floatSample(sampleTs.UnixMilli())
+		val := createWriteRequest(b, 1000, samples, nil)
+		produceRecords(ctx, b, writeClient, kafkaTime, "1", testTopic, 0, val)
+		expSamples = append(expSamples, samples...)
+	}
+
+	bb, err := New(cfg, log.NewNopLogger(), prometheus.NewPedanticRegistry(), overrides)
+	require.NoError(b, err)
+	compactCalled := make(chan struct{}, 10)
+	bb.tsdbBuilder = func() builder {
+		testBuilder := newTestTSDBBuilder(cfg, overrides)
+		testBuilder.compactFunc = func() {
+			compactCalled <- struct{}{}
+		}
+		return testBuilder
+	}
+
+	// It was difficult to reliably get the time taken to consume the records and produce a block
+	// by using b.ResetTimer. So for now, printing the required time to analyse.
+	// TODO(codesome): figure out how we can get the required numbers without prints.
```

Comment:
To be noted about the benchmark. The 8% gains come from the printed time below.

Reply:
> figure out how we can get the required numbers without prints

You can use `b.ResetTimer` and [`b.ReportMetric`](https://pkg.go.dev/testing#B.ReportMetric) for that:

```go
bb := New()
bb.staring()

b.ResetTimer()

err := bb.NextConsumeCycle(ctx, cycleEnd)
if err != nil {
    b.Fatal(err)
}
<-compactCalled

b.ReportMetric(float64(b.Elapsed().Nanoseconds()), "ns/op")
```

Reply:
I tried the ResetTimer but the benchmark seemed to be running many iterations like forever. So I had removed it to make the benchmark timer go long. TIL about ReportMetric! I will try that